pkg/backup/remap_crd_version_action.go

```diff
@@ -0,0 +1,114 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	v1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+)
+
+// RemapCRDVersionAction inspects CustomResourceDefinition and decides if it is a v1
+// CRD that needs to be backed up as v1beta1.
+type RemapCRDVersionAction struct {
+	logger logrus.FieldLogger
+}
+
+// NewRemapCRDVersionAction instantiates a new RemapCRDVersionAction plugin.
+func NewRemapCRDVersionAction(logger logrus.FieldLogger) *RemapCRDVersionAction {
+	return &RemapCRDVersionAction{logger: logger}
+}
+
+// AppliesTo selects the resources the plugin should run against. In this case, CustomResourceDefinitions.
+func (a *RemapCRDVersionAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"customresourcedefinition.apiextensions.k8s.io"},
+	}, nil
+}
+
+// Execute executes logic necessary to check a CustomResourceDefinition and inspect it for characteristics that necessitate saving it as v1beta1 instead of v1.
+func (a *RemapCRDVersionAction) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	a.logger.Info("Executing RemapCRDVersionAction")
+
+	// This plugin is only relevant for CRDs retrieved from the v1 endpoint that were installed via the v1beta1
+	// endpoint, so we can exit immediately if the resource in question isn't v1.
+	apiVersion, ok, err := unstructured.NestedString(item.UnstructuredContent(), "apiVersion")
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to read apiVersion from CRD")
+	}
+	if ok && apiVersion != "apiextensions.k8s.io/v1" {
+		a.logger.Info("Exiting RemapCRDVersionAction, CRD is not v1")
+		return item, nil, nil
+	}
+
+	// We've got a v1 CRD, so proceed.
+	var crd apiextv1.CustomResourceDefinition
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &crd); err != nil {
+		return nil, nil, errors.Wrap(err, "unable to convert unstructured item to a v1 CRD")
+	}
+
+	log := a.logger.WithField("plugin", "RemapCRDVersionAction").WithField("CRD", crd.Name)
+
+	// Looking for 1 version should be enough to tell if it's a v1beta1 CRD, as all v1beta1 CRD versions share the same schema.
+	// v1 CRDs can have different schemas per version
+	// The silently upgraded versions will often have a `versions` entry that looks like this:
+	//   versions:
+	//   - name: v1
+	//     served:  true
+	//     storage: true
+	// This is acceptable when re-submitted to a v1beta1 endpoint on restore.
+	if len(crd.Spec.Versions) > 0 {
```

Comment:
Would this be a more accurate condition for this?
```suggestion
	if len(crd.Spec.Versions) == 1 {
```